File: 1999/3/6-3.py
#!/usr/bin/env python

import sys

class PTree:
    class Node:
        def __init__(self, d):
            self.data = d
            self.left = None
            self.right = None

    def __init__(self, line):
        self.isPalin = 0
        self.pal = ''
        self.root = self.create(line)
        s = ''
        self.traverseAll(self.root, s)

    def isPal(self):
        return self.isPalin

    def getPal(self):
        return self.pal

    def create(self, line):
        line.pop(0)
        ch = line.pop(0)
        if ch == ')':
            return None
        tmp = PTree.Node(ch)
        tmp.left = self.create(line)
        tmp.right = self.create(line)
        line.pop(0)
        return tmp

    def traverseAll(self, n, s):
        s = s + str(n.data)
        if n.left == None and n.right == None:
            self.isPalin = self.isPalindrome(s)
        if self.isPalin:
            self.pal = s
        if n.left != None and not self.isPalin:
            self.traverseAll(n.left, s)
        if n.right != None and not self.isPalin:
            self.traverseAll(n.right, s)

    def isPalindrome(self, s):
        for i in xrange(len(s) - 1):
            if s[i] != s[-i - 1]:
                return 0
        return 1

cnt = 1
for line in sys.stdin:
    line = list(line)
    t = PTree(line)
    print "Tree %s is" % cnt,
    if t.isPal():
        print "a palintree. Its palintree is %s." % t.getPal()
    else:
        print "not a palintree."
    cnt += 1

File: 1999/3/6-3.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;

class PTree
{
private:
	struct Node
	{
		char data;
		Node* left, * right;
		Node(char d) : data(d), right(NULL), left(NULL) {}
	};
	
	Node* root;
	bool isPalin;
	string pal;
	
	Node* create(ifstream& in)
	{
		Node* tmp;
		char ch, dummy;
		in >> dummy >> ch;
		if (ch == ')')
			return NULL;
		tmp = new Node(ch);
		tmp->left = create(in);
		tmp->right = create(in);
		in >> dummy;
		return tmp;
	}
	
	void traverseAll(Node* n, string s)
	{
		s.push_back(n->data);
		if ((n->left == NULL) && (n->right == NULL))
			isPalin = isPalindrome(s);
		if (isPalin)
			pal = s;
		if ((n->left != NULL) && !isPalin)
			traverseAll(n->left, s);
		if ((n->right != NULL) && !isPalin)
			traverseAll(n->right, s);
	}
	
	bool isPalindrome(string s)
	{
		for (unsigned i = 0; i < (s.size() - i); i++)
			if (s[i] != s[s.size() - i - 1])
				return false;
		return true;
	}

public:
	bool make(ifstream& in)
	{
		isPalin = false;
		root = create(in);
		string s;
		traverseAll(root, s);
		return true;
	}
	
	bool isPal()
	{
		return isPalin;
	}
	
	string getPal()
	{
		return pal;
	}
};

void main()
{
	ifstream in("prob3_in.txt");
	int cnt = 1;
	PTree t;
	while (!in.eof())
	{
		t.make(in);
		cout << "Tree " << cnt << " is ";
		if (t.isPal())
			cout << "a palintree. Its palindrome is " << t.getPal() << "." << endl;
		else
			cout << "not a palintree." << endl;
		cnt++;
	}
}
